packages/playwright-core/src/server/injected/roleUtils.ts

```diff
@@ -461,6 +461,35 @@ export function getElementAccessibleDescription(element: Element, includeHidden:
   return accessibleDescription;
 }
 
+export function getElementAccessibleErrorMessage(element: Element): string {
+  const cache = cacheAccessibleErrorMessage;
+  let accessibleErrorMessage = cacheAccessibleErrorMessage?.get(element);
+
+  if (accessibleErrorMessage === undefined) {
+    accessibleErrorMessage = '';
+
+    const ariaInvalid = element.getAttribute('aria-invalid');
+    const isAriaInvalid = ariaInvalid !== null && ariaInvalid.toLowerCase() !== 'false';
```

Comment:
- According to [the spec](https://w3c.github.io/aria/#aria-invalid), empty value is also `false`.
- Let's introduce `getAriaInvalid()` helper similar to `getAriaLevel()`.
- Let's check [HTMLInputElement.validity.valid](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/validity) in addition to `aria-invalid`.
- Let's check the `getAriaRole()` to be the one explicitly listed [in the spec](https://w3c.github.io/aria/#aria-errormessage) to future proof for Aria 1.2+.

Reply:
Thanks for nice direction and feedback. :)

---
 I have researched the relationship between `input validity` and `aria-invalid` but could not find documentation explicitly addressing their priority.

As a result, I implemented the following approach: for `HTMLInputElement` elements, if attributes affecting `input validity` are present, the code prioritizes checking the `validity` property. Below is the implemented code. I would appreciate your feedback on [this approach](https://github.com/microsoft/playwright/pull/33904/commits/3c01c81ab3915b0c0954e9b0c481ac2953e0886e).

```ts
export function getAriaInvalid(element: Element): 'false' | 'true' | 'grammar' | 'spelling' {
  const role = getAriaRole(element) || '';
  if (!role || !kAriaInvalidRoles.includes(role))
    return 'false';
  // ⛳️
  if (element instanceof HTMLInputElement && hasValidationAttributes(element) && element.validity)
    return element.validity.valid ? 'false' : 'true';
  const ariaInvalid = element.getAttribute('aria-invalid');
  if (!ariaInvalid || ariaInvalid.trim() === '' || ariaInvalid.toLocaleLowerCase() === 'false')
    return 'false';
  if (ariaInvalid === 'true' || ariaInvalid === 'grammar' || ariaInvalid === 'spelling')
    return ariaInvalid;
  return 'true';
}
```